android/filament-utils-android/src/main/java/com/google/android/filament/utils/ModelViewer.kt

```diff
@@ -0,0 +1,268 @@
+    }
+
+    @Suppress("unused")
+    constructor(textureView: TextureView) {
+        cameraManipulator = Manipulator.Builder()
+                .targetPosition(0.0f, 0.0f, -4.0f)
+                .viewport(textureView.width, textureView.height)
+                .build(Manipulator.Mode.ORBIT)
+
+        gestureDetector = GestureDetector(textureView, cameraManipulator)
+        uiHelper.renderCallback = SurfaceCallback()
+        uiHelper.attachTo(textureView)
+    }
+
+    /**
+     * Loads a monolithic binary glTF and populates the Filament scene.
+     */
+    fun loadModelGlb(buffer: Buffer) {
+        destroyModel()
+        asset = assetLoader.createAssetFromJson(buffer)
+        asset?.let { asset ->
+            val resourceLoader = ResourceLoader(engine)
+            resourceLoader.loadResources(asset)
+            resourceLoader.destroy()
+            animator = asset.animator
+            asset.releaseSourceData()
+            scene.addEntities(asset.entities)
+        }
+    }
+
+    /**
+     * Loads a JSON-style glTF file and populates the Filament scene.
+     */
+    fun loadModelGltf(buffer: Buffer, callback: (String) -> Buffer) {
+        destroyModel()
+        asset = assetLoader.createAssetFromJson(buffer)
+        asset?.let { asset ->
+            val resourceLoader = ResourceLoader(engine)
+            for (uri in asset.resourceUris) {
+                resourceLoader.addResourceData(uri, callback(uri))
+            }
+            resourceLoader.loadResources(asset)
+            resourceLoader.destroy()
+            animator = asset.animator
+            asset.releaseSourceData()
+            scene.addEntities(asset.entities)
+        }
+    }
+
+    /**
+     * Sets up a root transform on the current model to make it fit into the viewing frustum.
+     */
+    fun transformToUnitCube() {
+        asset?.let { asset ->
+            val tm = engine.transformManager
+            val center = asset.boundingBox.center.let { v-> Float3(v[0], v[1], v[2]) }
+            val halfExtent = asset.boundingBox.halfExtent.let { v-> Float3(v[0], v[1], v[2]) }
+            val maxExtent = 2.0f * max(halfExtent)
+            val scaleFactor = 2.0f / maxExtent
+            center.z = center.z + 4.0f / scaleFactor
+            val transform = scale(Float3(scaleFactor)) * translation(Float3(-center))
+            tm.setTransform(tm.getInstance(asset.root), transpose(transform).toFloatArray())
+        }
+    }
+
+    /**
+     * Frees all entities associated with the most recently-loaded model.
+     *
+     * @see detach
+     */
+    fun destroyModel() {
+        asset?.let { asset ->
+            assetLoader.destroyAsset(asset)
+            this.asset = null
+            this.animator = null
+        }
+    }
+
+    /**
+     * Destroys the Filament engine and all related objects.
+     *
+     * The ModelViewer becomes invalid after this is called.
+     *
+     * @see destroyModel
+     */
+    fun detach() {
```

Comment:
There's a better way. You can set a `View.OnAttachStateChangeListener` to the `Surface/TextureView` at construction time and do all this automatically from `onViewDetachedFromWindow(View)`. See https://developer.android.com/reference/android/view/View.OnAttachStateChangeListener and https://developer.android.com/reference/android/view/View.html#addOnAttachStateChangeListener(android.view.View.OnAttachStateChangeListener)

Reply:
Done.